### PR TITLE
Configurable number of download thread.

### LINKED ISSUE #108: It is recommended to support configurable download threads
类似#82，线程过多会导致下载失败，建议改成可配置线程数。大概改了下，测了下爱奇艺可以
`	var wg sync.WaitGroup
	ParallelNum := 0
	// multiple fragments
	parts := []string{}
	for index, url := range data.URLs {
		partFileName := fmt.Sprintf("%s[%d]", title, index)
		partFilePath := utils.FilePath(partFileName, url.Ext, false)
		parts = append(parts, partFilePath)		

		if(ParallelNum + 1 > config.ParallelNum){			
			wg.Wait()
			ParallelNum = 0
		}

		ParallelNum = ParallelNum + 1		
		wg.Add(1)
		go func(url URLData, refer, fileName string, bar *pb.ProgressBar) {
			defer wg.Done()
			data.urlSave(url, refer, fileName, bar)
		}(url, refer, partFileName, bar)			
	}
	wg.Wait()
	bar.Finish()
`

### PR DESCRIPTION
**Pull request by cxjava**
close #108 

BTW, I removed the mgtv code, because this part of the code does not need.

```go
if strings.Contains(refer, "mgtv") {
        // Too many threads cause mgtv to return HTTP 403 error
       data.urlSave(url, refer, partFileName, bar)
}
```

### CODE REVIEW

File: utils/pool.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package utils
+
+import (
+	"math"
+	"sync"
+)
+
+type WaitGroupPool struct {
```
Comment (iawia002) [inline utils/pool.go]: the exported function should have comments (all the following functions need comments)

File: main.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ func init() {
 	flag.StringVar(&config.OutputPath, "o", "", "Specify the output path")
 	flag.StringVar(&config.OutputName, "O", "", "Specify the output file name")
 	flag.BoolVar(&config.ExtractedData, "j", false, "Print extracted data")
+	flag.IntVar(&config.ThreadNumber, "n", 100, "The number of download thread")
```
Comment (iawia002) [inline main.go]: I think `10` is enough

File: README.md [hunk 3]
```diff
@@ -317,6 +317,8 @@ Usage of annie:
     	Select specific format to download
   -i	Information only
   -j	Print extracted data
+  -n int
+        The number of download thread (default 100)
```
Comment (iawia002) [inline README.md]: this need update too (`default 10`)

File: utils/pool_test.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	"sync/atomic"
+	"testing"
+)
+
+func TestWaitGroupPool(t *testing.T) {
+	wgp := NewWaitGroupPool(10)
+
+	var total uint32
+
+	for i := 0; i < 1000; i++ {
```
Comment (iawia002) [inline utils/pool_test.go]: maybe 100? Although go is fast, we don't need unnecessary loop

**Comment by iawia002** [comment]
I will do the document work after this gets merged

File: utils/pool_test.go (hunk 4 above)
Comment (cxjava) [reply utils/pool_test.go]: done!

File: README.md (hunk 3 above)
Comment (cxjava) [reply README.md]: done!

File: main.go (hunk 2 above)
Comment (cxjava) [reply main.go]: done!

File: utils/pool.go (hunk 1 above)
Comment (cxjava) [reply utils/pool.go]: done!

**Comment by cxjava** [comment]
> By the way, you need to pull the latest code, modify it on that basis, I fixed the failed test just now.

I will create a new branch base on the newest code to create pull request next time.

**Comment by iawia002** [comment]
> I will create a new branch base on the newest code to create pull request next time.

yeah, that's the right way

Thanks again @cxjava

**Pull request merged**